docs/contributing/code-quality-principles.md

```diff
@@ -0,0 +1,167 @@
+We use constructor DI only.
+
+```diff
+class LocalizedType
+{
+    /**
+     * @var \Shopsys\FrameworkBundle\Model\Localization\Localization
+     */
+    private $localization;
+
+-   public function __construct()
+-   {
+-       $this->localization = new Localization();
+-   }
++   /**
++   * @param \Shopsys\FrameworkBundle\Model\Localization\Localization $localization
++   */
++   public function __construct(Localization $localization)
++   {
++       $this->localization = $localization;
++   }
+
+    // ...
+}
+```
+
+### Reasons
+
+It may be difficult to create the dependency - it may have also dependencies or configuration.
+Once creating the dependency is not the class responsibility, the class code is much more straight-forward.
+
+We inject only dependencies that the class needs, it is forbidden to pass the service locator to the class.
+With service locator injection it is difficult to understand what the class really do (or need) and such class is difficult to test.
+
+We use constructor DI so the class is always in a consistent state.
+
+It is essential to provide code extensibility in the framework and this is achieved with proper DI.
+
+### Exceptions
+
+We don't respect this principle in tests as it is impossible to pass dependencies to PHPUnit tests via the constructor.
+In tests we use `$this->getContainer()->get(...)`.
+
+## Strict types
+
+`strict_types` is a directive that enforces PHP to check the type of scalar types (`int`, `float`, `string`, `bool`) strictly.
+
+```diff
++ declare(strict_types=1);
+
+final class Money implements JsonSerializable
+{
+    // ...
+
+    public static function createFromFloat(float $float, int $scale): self
+    // ...
+}
+```
+
+If the caller of `createFromFloat()` method provides different arguments than declared, PHP throws a `TypeError`.
+This leads to safer code.
+
+### Reasons
+
+We use scalar types (`int`, `float`, `string`, `bool`) as much as possible in new code.
+Using scalar types without strict mode leads to automatic type cast that can be even harmful (eg. input string `""` is cast to integer `0`).
+We prevent this risky automatic type cast by declaring strict types.
+
+Please read more about scalar types and strict mode in [Scalar Type Declarations](https://wiki.php.net/rfc/scalar_type_hints_v5).
+
+## Don't repeat yourself
+
+> Every piece of knowledge must have a single, unambiguous, authoritative representation within a system.
+
+DRY is one of the most well-known programming principles and it was already written so much about this principle that it doesn't make sense to write it once again.
+Please find more about DRY in online sources or in the great book [The Pragmatic Programmer](https://pragprog.com/book/tpp/the-pragmatic-programmer) where this principle was formulated for the first time.
+
+### Exceptions
+
+We understand the DRY principle in a pragmatic way, we aren't obsessed when there is a duplication in the system that makes sense.
```

Comment:
```suggestion
We understand the DRY principle in a pragmatic way, we aren't obsessed with it when there is a duplication in the system that makes sense.
```

Reply:
I tried grammar and translator and end up with maybe even better formulation `...we aren't obsessed with a duplication...`